nifi-nar-bundles/nifi-registry-bundle/nifi-registry-service/src/main/java/org/apache/nifi/schemaregistry/services/InMemoryJsonSchemaRegistry.java

```diff
@@ -0,0 +1,156 @@
+import static org.apache.commons.lang3.StringUtils.isNotBlank;
+
+@Tags({"schema", "registry", "json"})
+@CapabilityDescription("Provides a service for registering and accessing schemas. One can register a schema "
+        + "as a dynamic property where 'name' represents the schema name and 'value' represents the textual "
+        + "representation of the actual schema following the syntax and semantics of JSON's Schema format. "
+        + "Empty schemas and schemas only consisting of whitespace are not acceptable schemas."
+        + "The registry is heterogeneous registry as it can store schemas of multiple schema draft versions. "
+        + "By default the registry is configured to store schemas of Draft 2020-12. When a schema is added, the version "
+        + "which is currently is set, is what the schema is saved as.")
+@DynamicProperty(name = "Schema name", value = "Schema Content",
+        description = "Adds a named schema using the JSON string representation of a JSON schema",
+        expressionLanguageScope = ExpressionLanguageScope.NONE)
+public class InMemoryJsonSchemaRegistry extends AbstractControllerService implements JsonSchemaRegistry, JsonSchemaRegistryComponent {
+
+    private static final List<PropertyDescriptor> PROPERTY_DESCRIPTORS = Collections.singletonList(SCHEMA_VERSION);
+
+    private final ConcurrentMap<String, JsonSchema> jsonSchemas;
+    private final ConcurrentMap<SchemaVersion, JsonSchemaFactory> schemaFactories;
+    private volatile SchemaVersion schemaVersion;
+
+    public InMemoryJsonSchemaRegistry() {
+        jsonSchemas = new ConcurrentHashMap<>();
+        schemaFactories = Arrays.stream(SchemaVersion.values())
+                .collect(Collectors.toConcurrentMap(Function.identity(),
+                        schemaDraftVersion -> JsonSchemaFactory.getInstance(SpecVersion.VersionFlag.fromId(schemaDraftVersion.getUri()).get())));
+        schemaVersion = SchemaVersion.valueOf(SCHEMA_VERSION.getDefaultValue());
+    }
+
+    @Override
+    public void onPropertyModified(final PropertyDescriptor descriptor, final String oldValue, final String newValue) {
+        if (SCHEMA_VERSION.getName().equals(descriptor.getName()) && !newValue.equals(oldValue)) {
+            schemaVersion = SchemaVersion.valueOf(newValue);
+        } else if(descriptor.isDynamic() && isBlank(newValue)) {
+            jsonSchemas.remove(descriptor.getName());
+        } else if (descriptor.isDynamic() && isNotBlank(newValue)) {
+            try {
+                final String schemaName = descriptor.getName();
+                final JsonSchemaFactory jsonSchemaFactory = schemaFactories.get(schemaVersion);
+                jsonSchemaFactory.getSchema(newValue);
+                jsonSchemas.put(schemaName, new JsonSchema(schemaVersion, newValue));
+            } catch (final Exception e) {
+                // not a problem - the service won't be valid and the validation message will indicate what is wrong.
```

Comment:
I can but I essentially copied this style from the `AvroSchemaRegistry` whose code is from lines 83-98
```
if(descriptor.isDynamic()) {
            // Dynamic property = schema, validate it
            if (newValue == null) {
                recordSchemas.remove(descriptor.getName());
            } else {
                try {
                    // Use a non-strict parser here, a strict parse can be done (if specified) in customValidate().
                    final Schema avroSchema = new Schema.Parser().setValidate(false).parse(newValue);
                    final SchemaIdentifier schemaId = SchemaIdentifier.builder().name(descriptor.getName()).build();
                    final RecordSchema recordSchema = AvroTypeUtil.createSchema(avroSchema, newValue, schemaId);
                    recordSchemas.put(descriptor.getName(), recordSchema);
                } catch (final Exception e) {
                    // not a problem - the service won't be valid and the validation message will indicate what is wrong.
                }
            }
        }
```

Reply:
So if call `getLogger` in the `catch` block I foresee an issue in the unit test as getLogger will return null.  How can I configure this controller service in the unit test with a mock logger?

Reply:
The Logger should not be null in this unit test if the Controller Service is enabled using the TestRunner. I did not note that detail one way or the other on previous reviews, but if that needs to be changed, this would be a good reason to do so.

Reply:
I do not think I can use the TestRunner to enable the controller service as it will always fail if there are no schemas configured. I am trying to test the schema retrieval logic in the unit test `TestInMemoryJsonSchemaRegistry#testSchemaRetrieval` on lines 96-108.

Reply:
Thanks for setting me straight. I got it working with adding and not enabling the controller service.